### DISCUSSION TITLE
ROS launch commands not working , autodrive_cpp folder does not build

### DISCUSSION THREAD

**Discussion by SumeetShanbhag** [Tinker-Twins/AutoDRIVE#6] [Q&A] [answered]
Hello guys, I was struggling with some of the following issues

1. The ROS commands that I mentioned seem to work now but when I try to launch multiple ROSlaunch commands in different terminals, I am unable to do so. I am unable to understand why this happens.

2. Some of the ROS launch commands do not seem to be working at all such as hector_slam and rviz ones. 

3. I was unable to build the autodrive_cpp folder and run into errors when I am trying to do that.

4. I was also curious to know after the connection is established between the simulator and ros_bridge, how to make the vehicle navigate in the map properly.

**Comment by Tinker-Twins** [comment, marked as answer]
Hey Sumeet,

Please find numbered responses to your issues/questions:

1. Certain commands (e.g. those for headless and RViz mode bringup) are either/or type launch commands and running one after the other will result in conflicts ending up killing some commonly invoked nodes/processes. The [documentation](https://github.com/Tinker-Twins/AutoDRIVE/blob/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_ros/README.md) has now been updated to reflect this more clearly.
2. Please note that in order to be able to successfully execute certain launch files such as `hector_slam`, you will need to install additional dependencies (e.g. [`hector_slam`](http://wiki.ros.org/hector_slam) ROS packages in this case). All the necessary ROS package dependencies have now been added to the [documentation](https://github.com/Tinker-Twins/AutoDRIVE/blob/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_ros/README.md).
3. Please note that the [C++ API](https://github.com/Tinker-Twins/AutoDRIVE/tree/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_cpp) requires certain dependencies without which you may encounter errors. We have a convenient [shell script](https://github.com/Tinker-Twins/AutoDRIVE/blob/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_cpp/build.sh) that we have tested with Ubuntu 18.04, which should automatically download and install all the required dependencies and build the [C++ API](https://github.com/Tinker-Twins/AutoDRIVE/tree/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_cpp) for you (alternatively you could also dig through the [shell script](https://github.com/Tinker-Twins/AutoDRIVE/blob/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_cpp/build.sh) yourself and install everything manually). That being said, we are currently working on testing the [C++ API](https://github.com/Tinker-Twins/AutoDRIVE/tree/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_cpp) with all-inclusive installation so that users would not need to worry about external dependencies! Please stay tuned to here from us about this in the near future.
4. In order to be able to make the vehicle (Nigel) navigate autonomously, you will need to:
      - Build map of the environment (e.g. using [`hector_slam`](http://wiki.ros.org/hector_slam)).
      - Save the built map (e.g. using [`map_server`](http://wiki.ros.org/map_server)).
      - Setup map-based localization (e.g. using [`amcl`](http://wiki.ros.org/amcl)).
      - Setup and configure navigation stack (e.g. using [`navigation`](https://wiki.ros.org/navigation))
      - Setup and configure planning and control modules (e.g. using [`teb_local_planner`](http://wiki.ros.org/teb_local_planner) and [`nav_ctrl`](https://github.com/Tinker-Twins/AutoDRIVE/blob/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_ros/autodrive/scripts/nav_ctrl.py))
      **Note:** All of these modules have already been implemented and tested using [`autodrive_ros`](https://github.com/Tinker-Twins/AutoDRIVE/tree/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_ros) package and all you really need to do is (build and) load a map of the environment, give initial pose estimate and publish a goal pose for the vehicle to attain. You can very well tinker around with the parameters associated with any/all of these algorithms and see how they affect the overall performance!

**Reply by SumeetShanbhag** [reply]
Thankyou so much. This was extremely helpful!

**Comment by Tinker-Twins** [comment]
> 3. I was unable to build the autodrive_cpp folder and run into errors when I am trying to do that.

We have released an update for [AutoDRIVE Devkit 0.2.2](https://github.com/Tinker-Twins/AutoDRIVE/releases/tag/Devkit-0.2.2) with an aim of improving the dependency management and installation of the C++ API. Kindly try installing the updated version using the instructions given in [`README.md`](https://github.com/Tinker-Twins/AutoDRIVE/blob/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_cpp/README.md) and let us know how it goes.

**Comment by SumeetShanbhag** [comment]
Thank you for letting me know. I will get back to you once I try it.
________________________________
From: Tinker Twins ***@***.***>
Sent: Monday, July 3, 2023 11:13 PM
To: Tinker-Twins/AutoDRIVE ***@***.***>
Cc: Shanbhag, Sumeet Dinesh ***@***.***>; Author ***@***.***>
Subject: [EXT] Re: [Tinker-Twins/AutoDRIVE] ROS launch commands not working , autodrive_cpp folder does not build (Discussion #6)

You don't often get email from ***@***.*** Learn why this is important<https://aka.ms/LearnAboutSenderIdentification>

  1.  I was unable to build the autodrive_cpp folder and run into errors when I am trying to do that.

We have released an update for AutoDRIVE Devkit 0.2.2<https://github.com/Tinker-Twins/AutoDRIVE/releases/tag/Devkit-0.2.2> with an aim of improving the dependency management and installation of the C++ API. Kindly try installing the updated version using the instructions given in README.md<https://github.com/Tinker-Twins/AutoDRIVE/blob/AutoDRIVE-Devkit/ADSS%20Toolkit/autodrive_cpp/README.md> and let us know how it goes.
